drift/views/v1.py

```diff
@@ -194,47 +196,71 @@ def comparison_report(
+            with app.app_context():
+                # can raise RBACDenied exception
+                message = "reading historical system profiles"
+                current_app.logger.audit(message, request=request)
+                hsp_results = fetch_historical_sys_profiles(
+                    historical_sys_profile_ids,
+                    auth_key,
+                    logger,
+                    counters,
+                )
+        return {"hsp_results": hsp_results}
+
     try:
-        systems_with_profiles = []
-        baseline_results = []
-        hsp_results = []
+        api_results = {}
 
         with PT_CR_API_REQUESTS.time():
-            try:
-                with PT_CR_API_SYSTEM_PROFILE_REQUESTS.time():
-                    if system_ids:
-                        # can raise RBACDenied exception
-                        message = "reading systems with profiles"
-                        current_app.logger.audit(message, request=request)
-                        systems_with_profiles = fetch_systems_with_profiles(
-                            system_ids, auth_key, current_app.logger, get_event_counters()
-                        )
-
-                with PT_CR_API_BASELINE_REQUESTS.time():
-                    if baseline_ids:
-                        # can raise RBACDenied exception
-                        message = "reading baselines"
-                        current_app.logger.audit(message, request=request)
-                        baseline_results = fetch_baselines(
-                            baseline_ids, auth_key, current_app.logger
-                        )
-                        ensure_correct_system_count(baseline_ids, baseline_results)
-
-                with PT_CR_API_HSP_REQUESTS.time():
-                    if historical_sys_profile_ids:
-                        # can raise RBACDenied exception
-                        message = "reading historical system profiles"
-                        current_app.logger.audit(message, request=request)
-                        hsp_results = fetch_historical_sys_profiles(
-                            historical_sys_profile_ids,
-                            auth_key,
-                            current_app.logger,
-                            get_event_counters(),
-                        )
-            except RBACDenied as error:
-                message = error.message
-                current_app.logger.audit(str(HTTPStatus.FORBIDDEN) + " " + message, request=request)
-                raise HTTPError(HTTPStatus.FORBIDDEN, message=message)
+            with concurrent.futures.ThreadPoolExecutor() as executor:
+                futures = []
+                app = current_app._get_current_object()
+                if system_ids:
+                    # can raise RBACDenied exception
+                    futures.append(executor.submit(get_systems_with_profiles, app, PT_CR_API_SYSTEM_PROFILE_REQUESTS, system_ids, auth_key, current_app.logger, get_event_counters()))
```

Comment:
You can swap the array with dict and use keys as the Future and a value noting which api result it represents.